lang/LangPrimSource/PyrListPrim.cpp

```diff
@@ -492,41 +492,98 @@ int prEvent_Delta(struct VMGlobals *g, int numArgsPushed)
 	PyrSlot *a, key, dur, stretch, delta;
 	double fdur, fstretch;
 	int err;
+	PyrClass *restClass = s_rest->u.classobj;
+	PyrSlot *slot;
 
 	a = g->sp;  // dict
 
 	SetSymbol(&key, s_delta);
 	identDict_lookup(slotRawObject(a), &key, calcHash(&key), &delta);
 
 	if (NotNil(&delta)) {
-		slotCopy(a,&delta);
+		if (isKindOfSlot(&delta, restClass)) {
+			slot = slotRawObject(&delta)->slots;
+			err = slotDoubleVal(slot, &fdur);
+		} else {
+			err = slotDoubleVal(&delta, &fdur);
+		};
+		if (err) {
+			return err;
+		} else {
+			SetFloat(g->sp, fdur);
+			return errNone;
+		}
 	} else {
 		SetSymbol(&key, s_dur);
 		identDict_lookup(slotRawObject(a), &key, calcHash(&key), &dur);
-
 		err = slotDoubleVal(&dur, &fdur);
 		if (err) {
-			if (NotNil(&dur)) return err;
-			SetNil(a);
-			return errNone;
+			if (NotNil(&dur)) {
```

Comment:
That is, the desired outcomes are:

- dur is a number: primitive returns the number.
- dur is nil: primitive returns nil.
- dur is a Rest: Check the rest's value.
  - If numeric, primitive returns the number.
  - If anything else, throw a wrong type error.
- dur is anything else: throw a wrong type error.

The only leaves with the same action are "Rest with non-numeric value" and "non-number, non-nil, non-Rest." But you have to do extra work to do to decide if the Rest's value is not numeric, and that can't be folded into one 'if' condition (at least, not without really appalling visual style).